Target generators (like `python_sources`) now only directly depend on their generated targets

stuhood opened the pull request:
Move (rather than copy) `Dependencies` fields from generators to generated targets. This avoids redundant graph edges and calculations for explicit dependencies. And in particular, it prevents us from needing to fill in missing parameters when computing the dependencies of generators (which is challenging, since they no longer have the `Field`s corresponding to their address values for comparison) for #14519.

This has a lot of fallout in tests, which in many cases were using "empty" generator targets (with no sources) as inner graph nodes. Because the target will no longer keep explicitly specified dependencies if it doesn't have sources, those targets either need to gain sources, or become generic `target`s.

To warn users of this behavior change, we additionally deprecate declaring a generator target with an explicitly empty `sources=[]` field.

[ci skip-build-wheels]

Eric-Arellano commented on src/python/pants/engine/internals/graph.py:
```diff
@@ -151,6 +154,27 @@ def warn_deprecated_field_type(field_type: type[Field]) -> None:
 
 
+@memoized
+def maybe_warn_dependencies_as_copied_field(tgt_type: type[TargetGenerator]) -> None:
```
Note that 2.11 already made a breaking change with target generators and moved vs copied fields. Didn't exist in 2.10. I don't think it's necessary to deprecate this particular change? Bundle it with the overall change.

stuhood replied:
This will convert to an error at some point, and there are outstanding PRs that will probably break. It's cheap enough to keep it as a deprecation.

stuhood commented:
This ended up larger than I would have hoped, since quite a few tests were relying on generator targets without sources preserving their (transitive) dependencies.

I expect that using a generator without sources as an alias for a batch of other targets is rare in production. But if that turns out not to be the case during the `2.11.0rc0`s, we can either 1) add a deprecation that recommends switching to `target` for those cases, 2) keep track of the explicit dependencies of a generator to use in case it doesn't have sources.

Eric-Arellano reviewed (commented):
Thanks!

> This has a lot of fallout in tests, which in many cases were using "empty" generator targets (with no sources) as inner graph nodes. Because the target will no longer keep explicitly specified dependencies if it doesn't have sources, those targets either need to gain sources, or become generic targets.

This doesn't seem like an internal change - that will impact users. Unlikely, hopefully, but possible. This should be an API change I think.

> Because the target will no longer keep explicitly specified dependencies if it doesn't have sources

But it does keep them if sources are specified? I don't understand that.

stuhood commented:
> > Because the target will no longer keep explicitly specified dependencies if it doesn't have sources
> 
> But it does keep them if sources are specified? I don't understand that.

They are moved to the sources. So the set of the things in the transitive graph doesn't change: it's just that the generator node _only_ depends on its sources, and nothing else.

Eric-Arellano commented:
I really like this change: I was playing with it and it looks the end-effect is that target generators now only directly depend on their generated targets, regardless of the `dependencies` field. I like that because it makes the semantics clearer of it being a generator, not an atom.

I'm concerned though that this is an API change, e.g. if you rely on the output of `./pants dependencies`. Ack that thanks to transitive deps, actual builds shouldn't change; but project introspection will.

Then there's the concern about depending on a target generator w/o `sources`.

> add a deprecation that recommends switching to target for those cases

What would this look like? Our transitive targets (and coarsened targets?) rules detect if you depend on a target generator w/ no `sources`? 

--

I recommend wording the PR to be more user-facing:

> Target generators like `python_sources` now only directly depend on their generated targets

We will want the 2.11 release blog to call attention to this also.

stuhood commented:
> That generally makes sense to me, but does interfere with repos who like to eagerly have a target generator in every dir like this as future-proofing for adding future tests:
> 
> ```python
> python_sources()
> python_tests(name="tests")
> ```
> 
> Related: we already have `[GLOBAL].files_not_found_behavior`.

Is this something that people actually do though? As you point out: to do it, they'd either have to disable `files_not_found_behavior` warnings, or clear the sources list.

Given that we have `files_not_found_behavior` to catch the case where the list was not empty, but didn't match anything, maybe the deprecation could just be for targets with explicitly empty sources lists: `sources=[]`?

> I think the issue is more _depending on an empty target generator_, rather than necessarily defining one. And it's not technically wrong to depend on it...just, meaningless & probably not what you intended.

It does seem to me that defining one is weird. `tailor` isn't going to do it for you, for example. So someone would have had to automate creation of a bunch of empty targets.

Eric-Arellano commented:
> Is this something that people actually do though?

I remember people doing this when we were helping them in the early days of Pants 2. Even pantsbuild/pants developers did it from bad copy-pasta. Not sure if post-tailor people still are doing it though.

> maybe the deprecation could just be for generators with explicitly empty sources lists: sources=[]?

I think that makes sense to me. 

One of my first thoughts was about how you could use an empty `python_sources` target to force Pants to use certain interpreter constraints. We actually used to tell people to do that when `pex_binary` temporarily did not have `interpreter_constraints`! But I see now what the impact of `interpreter_constraints` being a moved field would entail, and I think making it a moved field is 100% the right thing to do. So at that point, no reason to not use `target()` instead of `python_sources`.

--

Btw, I think my bigger concern with landing this PR is project introspection. But I don't see a way to gracefully deprecate that, like we could not with Pants 2.8. So instead, I think renaming the PR + mentioning in release blog will suffice.

stuhood commented:
Have updated the title, added a deprecation for explicitly empty sources lists on `TargetFilesGenerator` targets, and updated the description.

There will be a bunch more tests that need updating for this deprecation, some of which might uncover actual behavior changes or cases where use of `target` needs to be better supported. But I'd like to do that in a followup, due to the amount of test churn already landing here.

kaos commented on src/python/pants/backend/project_info/peek.py:
```diff
@@ -124,17 +121,17 @@ async def get_target_data(
+    expanded_dependencies = [
+        tuple(dep.address.spec for dep in deps)
+        for tgt, deps in zip(sorted_targets, dependencies_per_target)
+    ]
```
@stuhood `tgt` is unused in this comprehension, so no need to `zip` in `sorted_targets` (or is there another mistake hiding in here)?
